### PR TITLE
Re-factored how persistence data is generated + introduced a new way to do it for automated checks

### PR DESCRIPTION
**Pull request by waabid**
#### Pull request checklist

- [N/A] Addresses an existing issue: Fixes #0000
- [x] Added relevant unit test for your changes. (`npm run test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`npm run precheckin`)
- [n/a] Added screenshots/GIFs for UI changes.

#### Description of changes

Re-factored how persistence data is generated by splitting into separate modules and also introduced a new way, which specifically checks to ensure all requirements have persisted data for a test and that they have been scanned before trying to persist the data (useful in the automated check scenario).

**Please note that this new capability is not being used currently since it would cause test breakages due to a direct dependency in assessment store, which will be handled in another PR.**

#### Notes for reviewers

I believe we can re-factor even more and _modularize_ the persistence generation code but I did not feel that would be good to do right now.

### CODE REVIEW

File: src/background/create-initial-assessment-test-data.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    IGeneratedAssessmentInstance,
+    IManualTestStepResult,
+    InstanceIdToInstanceDataMap,
+    RequirementIdToResultMap,
+} from '../common/types/store-data/iassessment-result-data';
+import { DictionaryStringTo } from '../types/common-types';
+
+export type InitialDataCreator = (test: Readonly<Assessment>, persistedTest: IAssessmentData) => IAssessmentData;
+
+export const createInitialAssesmentTestData: InitialDataCreator = (test: Readonly<Assessment>, persistedTest: IAssessmentData) => {
```
Comment (dbjorge) [inline src/background/create-initial-assessment-test-data.ts]: `createInitialAssesmentTestData`: misspelling on Assesment

File: src/background/create-initial-assessment-test-data.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+): IAssessmentData {
+    const testData: IAssessmentData = getDefaultTestResult();
+    testData.testStepStatus = constructRequirementStatus(requirements, persistedRequirementsStatus);
+    testData.manualTestStepResultMap = constructManualRequirementResultMap(requirements, persistedManualMap);
+    testData.generatedAssessmentInstancesMap = constructGeneratedAssessmentInstancesMap(requirements, persistedGeneratedMap);
+    return testData;
+}
+
+function allRequirementsAreScanned(requirements: string[], persistedTest: IAssessmentData): boolean {
+    return !requirements.some(requirement => {
```
Comment (dbjorge) [inline src/background/create-initial-assessment-test-data.ts]: I think this impl would read more clearly if you applied DeMorgan's Law

File: src/background/create-initial-assessment-test-data.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+): IAssessmentData {
+    const testData: IAssessmentData = getDefaultTestResult();
+    testData.testStepStatus = constructRequirementStatus(requirements, persistedRequirementsStatus);
+    testData.manualTestStepResultMap = constructManualRequirementResultMap(requirements, persistedManualMap);
+    testData.generatedAssessmentInstancesMap = constructGeneratedAssessmentInstancesMap(requirements, persistedGeneratedMap);
+    return testData;
+}
+
+function allRequirementsAreScanned(requirements: string[], persistedTest: IAssessmentData): boolean {
+    return requirements.every(requirement => {
```
Comment (peterdur) [inline src/background/create-initial-assessment-test-data.ts]: Maybe a little less wordy as
```
    return requirements.every(
        requirement => persistedTest.testStepStatus[requirement] && persistedTest.testStepStatus[requirement].isStepScanned === true,
    );
```

File: src/tests/unit/tests/background/initial-assessment-store-data-generator.test.ts [hunk 4]
```diff
@@ -1,32 +1,29 @@
...
+import { IAssessmentStoreData } from '../../../../common/types/store-data/iassessment-result-data';
 import { VisualizationType } from '../../../../common/types/visualization-type';
 import { CreateTestAssessmentProvider } from '../../common/test-assessment-provider';
 
 describe('InitialAssessmentStoreDataGenerator.generateInitialState', () => {
     const assesssmentsProvider = CreateTestAssessmentProvider();
     const generator = new InitialAssessmentStoreDataGenerator(assesssmentsProvider);
 
-    const defaultState = generator.generateInitialState();
+    let defaultState;
```
Comment (smoralesd) [inline src/tests/unit/tests/background/initial-assessment-store-data-generator.test.ts]: does this not throw a tslint error for not having a type?

**Pull request merged**